extensions/lifecycle/util/RulesReducer.js

```diff
@@ -0,0 +1,282 @@
+     * NOTE: On a versioned bucket, three types of listing can be performed:
+     * - "current" targeting current objects (Expiration and Transitions).
+     * - "noncurrent" targeting non-current obejcts (NonCurrentExpiration and NonCurrentTransitions).
+     * - "orphan" targeting orphan delete markers (Expiration and Expiration.ExpiredObjectDeleteMarker).
+     * @return {object} aggregatedRules - aggregated rules for each type
+     * @return {array} aggregatedRules.currents - array of rules
+     * @return {array} aggregatedRules.nonCurrents - array of rules
+     * @return {array} aggregatedRules.orphans - array of rules
+     */
+    _getListingsForVersionedBucket() {
+        const reducedRules = this._bucketLCRules.reduce((accumulator, r) => {
+            const nonCurrents = accumulator.nonCurrents;
+            const currents = accumulator.currents;
+            const orphans = accumulator.orphans;
+
+            const reducedCurrents = this._reduceCurrentRules(r, currents);
+            const reducedNonCurrents = this._reduceNonCurrentRules(r, nonCurrents);
+            const reducedOrphans = this._reduceOrphanDeleteMarkerRule(r, orphans);
+
+            return { currents: reducedCurrents, nonCurrents: reducedNonCurrents, orphans: reducedOrphans };
+        }, { currents: [], nonCurrents: [], orphans: [] });
+
+        return reducedRules;
+    }
+
+    /**
+     * _reduceCurrentRules: checks if the rule targets current versions and, if so,
+     * adds its result to the listings array.
+     * The listings will be used to defined the parameters of "list current versions".
+     * @param {object} r - lifecycle rule to be evaluated
+     * @param {array} currents - array of current listings information
+     * @return {array} aggregatedListings - array of current listings information after evaluation
+     */
+    _reduceCurrentRules(r, currents) {
+        // handle the case when rule is disabled.
+        if (r.Status !== 'Enabled') {
+            return currents;
+        }
+        const prefix = this._getRulePrefix(r);
+        const isTransitions = r.Transitions && r.Transitions[0];
+        let days;
+
+        // IMPROVEMENT: ZENKO-4541 If only one transition (no expiration) rule is set for a given prefix,
+        // we can add the filter "value.dataStoreName != destinationLocationName" to our listing query
+        // to only return keys that have not been transitioned yet.
+
+        if (r.Expiration) {
+            // NOTE: Expiration Days cannot be 0.
+            if (r.Expiration.Days) {
+                days = this._decrementExpirationDay(r.Expiration.Days);
+            } else if (r.Expiration.Date) {
+                if (r.Expiration.Date <= this._currentDate) {
+                    days = 0;
+                }
+            }
+        }
+
+        if (isTransitions) {
+            // NOTE: Transitions Days cannot be 0.
```

Comment:
Not sure i understood this comment, seems it still works when having a rule set to `0 days`.
Also i believe pensieve accepts a value of 0 for current transitions
https://github.com/scality/pensieve-api/blob/b86dfec12c2d1a3e3d20d5a23dc46625f5ef7dee/gen/models/bucket_workflow_transition_v2_all_of1.go#L35-L37

Reply:
yes, good catch, my comment is wrong but the code logic should be right. (I am checking for `if (r.Transitions[0].Days !== undefined) {` )
My comment should have been `NOTE: Transitions Days can be 0`.